src/components/StatusPanel.jsx

```diff
@@ -0,0 +1,90 @@
+import React from 'react';
+import GlobalStatus from '../utils/globalStatus';
+import GlobalStatusSummary from './GlobalStatusSummary';
+import StatusHeader from './StatusHeader';
+import ServiceGrid from './ServiceGrid';
+import {
+  statusEndpoints, statuses, maintenanceWindows,
+} from '../config';
+import * as processors from '../utils/endpointParsers';
+import areBeingMaintained from '../utils/maintenanceUtils';
+
+class StatusPanel extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { endpoints: statusEndpoints };
+  }
+
+  componentDidMount() {
+    const { endpoints } = this.state;
+    const maintenanceWindow = areBeingMaintained(new Date(), maintenanceWindows);
+    const groupedEndpoints = this.groupedEndpoints();
+
+    Object.keys(groupedEndpoints).forEach((endpointUrl) => {
+      fetch(endpointUrl, {
+        mode: 'cors',
+      }).then((response) => {
+        Object.keys(groupedEndpoints[endpointUrl]).forEach((key) => {
+          if (response.status !== 200) {
+            const newState = endpoints;
+            if (maintenanceWindow) {
+              newState[key].status = 'maintenance';
+            } else {
+              newState[key].status = 'issue';
+            }
+            this.setState(newState);
+            return;
+          }
+
+          processors[endpoints[key].processor](response.clone())
+            .then((status) => {
+              const newState = endpoints;
+              newState[key].status = status;
+              this.setState(newState);
+            });
+        });
+      }).catch(() => {
+        Object.keys(groupedEndpoints[endpointUrl]).forEach((key) => {
+          const newState = endpoints;
+          newState[key].status = 'outage';
+          this.setState(newState);
+        });
+      });
+    });
+  }
+
+  groupedEndpoints() {
+    const { endpoints } = this.state;
+    const endpointCache = {};
+
+    Object.keys(endpoints).forEach((key) => {
+      const current = endpoints[key];
+      const { endpointUrl } = current;
+
+      if (endpointCache[endpointUrl]) {
+        endpointCache[endpointUrl][key] = current;
+      } else {
+        endpointCache[endpointUrl] = { [key]: current };
+      }
+    });
+    return endpointCache;
+  }
+
+  render() {
+    const { endpoints } = this.state;
+
+    return (
+      <>
+        <GlobalStatusSummary
+          status={new GlobalStatus(statuses, endpoints).status}
+        />
+        <StatusHeader
+          statuses={statuses}
+        />
+        <ServiceGrid endpoints={endpoints} />
```

Comment:
I think this is a big improvement for sure. I wish we could get out of the business of having to rerender `ServerGrid` on every status change.. but I'm not sure how that might go with the GlobalStatus business.

Reply:
Yeah, totally. I went down that path actually and came back up the chain to cleanup the higher level element and realized the coupling between the state and those two components.  Hopefully this will put us in a good position to change that (either soon, or down the road if we pick this up again).

Reply:
👍